### ISSUE TITLE
[LifetimeSafety] Detect invalidation of pointer parameters

### ISSUE THREAD

**Issue by usx95** [llvm/llvm-project#193044]
```cpp
#include <iostream>
#include <string>
#include <string_view>
#include <vector>

void fine() {
    std::vector<int> vv;
    auto it = vv.begin();
    vv.push_back(42); // detected. Ok.
    (void)it;
}

// Nothing detected below.
void foo() {
    std::vector<int> vv;
    std::vector<int>& v = vv;
    auto it = vv.begin();
    v.push_back(42);
    (void)it;
}

void bar() {
    std::vector<int> vv;
    std::vector<int>* v = &vv;
    auto it = vv.begin();
    v->push_back(42);
    (void)it;
}

void baz(std::vector<int>* v) {
    auto it = v->begin();
    v->push_back(42);
    (void)it;
}

```
https://godbolt.org/z/1j951a88E

**Comment by zeyi2** [comment]
Hi, I'd like to work on this :)

**Comment by usx95** [comment]
If you would prefer, I am happy to discuss any approaches here.

```cpp
// Some false-positives which might surface
void baz(std::vector<int>* v) {
    // auto it = v->begin();
    v->push_back(42);
    (void)v; // I think the analysis would not be able to differentiate 'v' and 'it' and this might be flagged as a UAI.
}
```

**Comment by zeyi2** [comment]
> If you would prefer, I am happy to discuss any approaches here.

Thanks, I would definitely appreciate discussing the approach before I turn this into a PR.

I tried reading the implementation of the current implementation, my understanding is that the missed diagnostics here come from the fact that we fail to model the invalidation event. (the heuristics is too conservative).

https://github.com/llvm/llvm-project/blob/ef867c03d77cc5c11845ab3ac2392c124271a9cf/clang/lib/Analysis/LifetimeSafety/FactsGenerator.cpp#L767-L779

What I tried on my side was to extend this modeling to a few additional aliasing shapes by matching AST Nodes, IMO this is a dirty approach since it is not alias-aware and hard to maintain/extend in the long run. I also tried to look for a more structural solution, but it seems that `AccessPath` is still root-only, which means I can't distinguish  `s.strings1.push_back(...)` and `s.strings2.push_back(...)`.

So from what I see, broadening invalidation modeling naively would likely introduce FPs, at the moment I seem to be stuck here (a more structural solution that seems to require more precise path modeling first). 

Is there another direction you would recommend here?

**Comment by usx95** [comment]
> I also tried to look for a more structural solution, but it seems that `AccessPath` is still root-only, which means I can't distinguish `s.strings1.push_back(...)` and `s.strings2.push_back(...)`.
> 
> So from what I see, broadening invalidation modeling naively would likely introduce FPs, at the moment I seem to be stuck here (a more structural solution that seems to require more precise path modeling first).

That's correct. Invalidations already has quite a few false-positives atm and all of them are because of this limitation. So additional FP should be fine for now. We can fix them in precise path modelling in a separate PR.

> > If you would prefer, I am happy to discuss any approaches here.
> 
> Thanks, I would definitely appreciate discussing the approach before I turn this into a PR.
> 
> I tried reading the implementation of the current implementation, my understanding is that the missed diagnostics here come from the fact that we fail to model the invalidation event. (the heuristics is too conservative).
> ```cpp
>  void FactsGenerator::handleInvalidatingCall(const Expr *Call, 
>                                              const FunctionDecl *FD, 
>                                              ArrayRef<const Expr *> Args) { 
>    const auto *MD = dyn_cast<CXXMethodDecl>(FD); 
>    if (!MD || !MD->isInstance()) 
>      return; 
>   
>    if (!isContainerInvalidationMethod(*MD)) 
>      return; 
>    // Heuristics to turn-down false positives. 
>    auto *DRE = dyn_cast<DeclRefExpr>(Args[0]); 
>    if (!DRE || DRE->getDecl()->getType()->isReferenceType()) 
>      return; 
> ```
> What I tried on my side was to extend this modeling to a few additional aliasing shapes by matching AST Nodes, IMO this is a dirty approach since it is not alias-aware and hard to maintain/extend in the long run. 

Happy to make this less conservative (instead of continued special casing). I think I added this precisely to get rid of some false-positives but this is not a long term solution.

**Comment by usx95** [comment]
(fix is reverted. potential prerequisite is https://github.com/llvm/llvm-project/issues/195706)
